src/dtree.ml

```diff
@@ -0,0 +1,499 @@
+  in
+  begin
+    match tree with
+    (* First step must be done to avoid drawing a top node. *)
+    | Node({ swap ; children = ch ; store ; default }) ->
+        F.fprintf ppf "@ 0 [label=\"%d\"%s];"
+          swap (if store then " shape=\"box\"" else "") ;
+      TcMap.iter (fun sw c -> write_tree 0 (DotCons(sw)) c) ch ;
+      (match default with None -> () | Some(tr) -> write_tree 0 DotDefa tr)
+    | Leaf(_)                                          -> ()
+    | _                                                -> assert false
+  end ;
+  F.fprintf ppf "@.}@\n@?" ;
+  close_out ochan
+
+(** {3 Clause matrix and pattern matching problem} *)
+
+(** A clause matrix encodes a pattern matching problem.  The clause matrix {i
+    C} can be denoted {i C = P → A} where {i P} is a {e pattern matrix} and {i
+    A} is a column of {e actions}.  Each line of a pattern matrix is a pattern
+    to which is attached an action.  When reducing a term, if a line filters
+    the term, or equivalently the term matches the pattern, the term is
+    rewritten to the action. *)
+module ClauseMat =
+struct
+  (** A component of the matrix, a term along with its position in the top
+      term. *)
+  type component = term * Subterm.t
+
+  (** [pp_component o c] prints component [c] to channel [o]. *)
+  let pp_component : component pp = fun oc (te, pos) ->
+    Format.fprintf oc "@[<h>%a@ %@@ %a@]" Print.pp te Subterm.pp pos
+
+  (** A redefinition of the rule type.
+
+      {b Note} that {!type:array} is used while {!module:ReductionStack} could
+      be used because {!val:pick_best_among} goes through all items of a rule
+      anyway ([max(S) = Θ(|S|)]).  Since heuristics need to access elements of
+      the matrix, we favour quick access with {!type:array}. *)
+  type rule =
+    { lhs : component array
+    (** Left hand side of a rule.   *)
+    ; rhs : action
+    (** Right hand side of a rule. *)
+    ; env_builder : (int * int) list
+    (** Maps an index int the {!val:vars} array to a slot. *) }
+
+  (** Type of a matrix of patterns.  Each line is a row having an attached
+      action. *)
+  type t =
+    { clauses : rule list
+    (** The rules. *)
+    ; var_met : int
+    (** Number of variables encountered so far. *) }
+
+  (** [pp o m] prints matrix [m] to out channel [o]. *)
+  let pp : t pp = fun oc { clauses ; _ } ->
+    let module F = Format in
+    let pp_line oc l =
+      F.fprintf oc "@[<h>" ;
+      F.pp_print_list ~pp_sep:(fun _ () -> F.fprintf oc ";@ ")
+        pp_component oc (Array.to_list l) ;
+      F.fprintf oc "@]" in
+    F.fprintf oc "{@[<v>@," ;
+    F.pp_print_list ~pp_sep:F.pp_print_cut pp_line oc
+      (List.map (fun { lhs ; _ } -> lhs) clauses) ;
+    F.fprintf oc "@.}@,"
+
+  (** [of_rules r] creates the initial pattern matrix from a list of rewriting
+      rules. *)
+  let of_rules : Terms.rule list -> t = fun rs ->
+    let r2r r =
+      let term_pos = List.to_seq r.Terms.lhs |> Subterm.tag |> Array.of_seq in
```

Comment:
Why is it interesting to use Seq here?

Reply:
I had written `tag` (it does not exist anymore in other branches) such that it took a seq as input, to operate either arrays or list without having to traverse to convert